docs/contributor/memory_management.md

```diff
@@ -0,0 +1,177 @@
+
+- If a value is stored somewhere, no `incRef`/`decRef` calls are needed (again, there are rare exceptions to this in the compiler, specifically regarding "swap slots", which are local variables used as registers).
+- If the computation no longer needs a live value (e.g. after a statement in the middle of a block), it should be `decRef`ed.
+- No special `incRef` is needed for return values, as the access of those values performs the requisite `incRef`. It should be noted that in generated WebAssembly code, the function postamble looks as follows:
+
+```
+swap_slot <- return_value
+for argument in arguments:
+    decref(argument)
+for local_variable in local_variables:
+    decref(local_variable)
+return load_swap(return_value) # see note below
+```
+
+Note that `load_swap` is one of the aforementioned exceptions which does not `incRef` the value upon access. Swap slots are used by the compiler to temporarily store intermediate values during computation, and they do not need to have their contents `incRef`ed; i.e. since they are only used for intermediate values, any references they hold are treated as weak and do not impact the reference count.
+
+**References are callee-owned.** This has a couple of implications:
+
+- If a function is called with an argument, it is the responsibility of that function to `decRef` that argument
+- If a value is passed to a function with a reference count of `n`, when it returns, it will have the a reference count of `n-1` (unless it was stored in an external location, in which case the reference count may be higher).
+
+**Functions contain references to themselves.** In order to support closures, Grain's compiler emits code which passes the current closure to each function as an extra argument. With respect to the calling convention, this extra argument is the same as normal arguments, so it follows the same `incRef`-before-function-call pattern as standard arguments (and must be `decRef`ed before functions return). This is particularly important to keep in mind when writing `@disableGC` code, as discussed below.
+
+## Disabling the Garbage Collector
+
+There are times in which the garbage collector is not desired in Grain code. The typical example of this is the implementation of the core runtime (for example the garbage collector itself), which involves working with native WebAssembly values. Consequently, Grain provides two mechanisms for disabling the garbage collector, either on a per-function or a per-file basis.
+
+First, it is possible to annotate functions with `@disableGC`. This will compile the annotated function with no garbage collection instructions. Second, it is possible to pass the `--compilation-mode=runtime` flag to the compiler (**NOTE** that this is not the only effect of this flag; see `runtime.md` for more details), which is typically done by placing this at the beginning of the file:
```

Comment:
This isn't quite right—the flag to turn off GC at the file level is `--no-gc`. `--compilation-mode=runtime` doesn't emit GC instructions more because the GC doesn't exist rather than it being disabled. We can probably just mention that code written in runtime mode that interacts with user code needs manual management.

If we wanted we could also include some information like this:
`--no-gc`: disables the garbage collector, allocations are still done via `malloc`
`--compilation-mode=runtime`: GC doesn't exist, implicit allocations can't be reclaimed